### PR TITLE
Add `LinesIter` algorithm to iterate over lines in a geometry

### LINKED ISSUE #708: Add `lines_iter` or `iter_lines` to `Polygon`
That iterates through all `Line`s in the exterior and interior rings of the `Polygon`

https://github.com/georust/geo/pull/701/files#r776821996

### PR DESCRIPTION
**Pull request by callpraths**
- [x] I agree to follow the project's [code of conduct](https://github.com/georust/geo/blob/master/CODE_OF_CONDUCT.md).
- [x] I added an entry to `CHANGES.md` if knowledge of this change could be valuable to users.
---

Newly introduced `LinesIter` is very similar to `CoordsIter`, but is implemented for only a subset of types (e.g., not for `Point`, and hence not for `Geometry`).

This PR also includes a small tweak to the order in which `Rect::to_lines()` returns the lines -- it now matches the winding order of the polygon created by `Rect::to_polygon()`.

Closes #708

### CODE REVIEW

File: geo/src/algorithm/lines_iter.rs
Comment (callpraths) [inline geo/src/algorithm/lines_iter.rs:92]: I would actually like to introduce newtypes for all these concrete `LinesIter::Iter` types because they currently leak implementation details (about how iterators are chained etc).

I followed the pattern in `CoordsIter` for this PR, but if there agreement, I'm happy to introduce newtypes for both `CoordsIter` and `LinesIter`.

File: geo/src/algorithm/lines_iter.rs
Comment (callpraths) [inline geo/src/algorithm/lines_iter.rs:79]: I am new(ish) to Rust, and I spent some time trying to wrap my head around this.

My current understanding is that as long as RFC-2071 is unstable, crate APIs should likely not use the `impl Trait` feature (in the return position, for existential types). Two reasons that I think I'd avoid this feature:
- Clients have no way to store the returned value in a struct / trait associated type (anywhere they must name the type).
- Conditional implementations (`#[derive(Debug)]` and friends) don't work well). In our case, `impl Iterator<Item=Line<T>` is not `Debug` even if both `T` and `Line` are).

Basing my understanding so far mostly off of https://www.reddit.com/r/rust/comments/8ik620/notes_on_impl_trait/

In this case, I was stuck because of the first issue -- I needed to name the type for `LinesIter::Iter`, but had no way to do it.

Comment (callpraths) [reply geo/src/algorithm/lines_iter.rs:79]: Did I say I'm new to Rust? Please let me know if I'm understanding this wrong / I'm missing something obvious in how `impl Trait` works.

Comment (callpraths) [reply geo/src/algorithm/lines_iter.rs:79]: Note to self: Update https://github.com/georust/geo/issues/681 in some form if `impl Trait` should be reconsidered in the public API of `geo-types`.

File: geo-types/CHANGES.md
Comment (lnicola) [inline geo-types/CHANGES.md:5]:
```suggestion
* BREAKING: Make `Rect::to_lines` return lines in winding order for `Rect::to_polygon`.
```

File: geo/src/algorithm/lines_iter.rs
Comment (lnicola) [inline geo/src/algorithm/lines_iter.rs:36]: Nit: I personally find these more distracting and not very useful, since they cover only one or two items.

File: geo/src/algorithm/lines_iter.rs
Comment (lnicola) [inline geo/src/algorithm/lines_iter.rs:62]: I believe `std` iterator structs aren't `#[doc(hidden)]`, and it's strange to use that for a type returned by the public API.

File: geo/src/algorithm/lines_iter.rs
Comment (lnicola) [inline geo/src/algorithm/lines_iter.rs:29]: Nit: these aren't formatted, but I don't think there's an easy way to fix that, so never mind.

File: geo/src/algorithm/lines_iter.rs
Comment (lnicola) [inline geo/src/algorithm/lines_iter.rs:81]:
```suggestion
            // SAFETY: slice::windows(2) is guaranteed to yield a slice with exactly 2 elements
```

File: geo/src/algorithm/lines_iter.rs
Comment (lnicola) [inline geo/src/algorithm/lines_iter.rs:148]: We might switch to 2021 in #741.

File: geo/src/algorithm/lines_iter.rs
Comment (lnicola) [inline geo/src/algorithm/lines_iter.rs:163]: Ditto.

**Review by lnicola — APPROVED** [review]

File: geo-types/CHANGES.md
Comment (callpraths) [reply geo-types/CHANGES.md]: Done.

File: geo/src/algorithm/lines_iter.rs
Comment (callpraths) [reply geo/src/algorithm/lines_iter.rs]: I agree they are not useful. I was trying to follow the convetion from `coords_iter`. Removed.

File: geo/src/algorithm/lines_iter.rs
Comment (callpraths) [reply geo/src/algorithm/lines_iter.rs]: Agreed. I had tried to follow what `coords_iter` does. Added a minimal docstring.

File: geo/src/algorithm/lines_iter.rs
Comment (callpraths) [reply geo/src/algorithm/lines_iter.rs]: Thanks for calling that out. Formatted.

My hack: Copy into an actual function -> format -> copy back (with a little help from editor to add/remove line comments)

File: geo/src/algorithm/lines_iter.rs
Comment (callpraths) [reply geo/src/algorithm/lines_iter.rs]: Done.

File: geo/src/algorithm/lines_iter.rs
Comment (callpraths) [reply geo/src/algorithm/lines_iter.rs:136]: Should I add a TODO in the comment referencing the PR? I can remove it once 741 is merged.

File: geo/src/algorithm/lines_iter.rs
Comment (frewsxcv) [inline geo/src/algorithm/lines_iter.rs:9]: What do you think about implementing `LinesIter` for `Point` and `MultiPoint` that return empty iterators? Then we could implement `LinesIter` for `GeometryCollection`. I don't feel strongly about it, just an idea.

Comment (michaelkirk) [reply geo/src/algorithm/lines_iter.rs:9]: It **would** be nice to have LinesIter on GeometryCollection, but I can't decide how surprising it would be for Point/MultiPoint...

My instinct is to hold off on including it for now. It would be a strictly additive change at some point if we wanted to add it.

File: geo-types/CHANGES.md
Comment (michaelkirk) [inline geo-types/CHANGES.md:5]: This was probably oversight. Good catch! 

I personally don't think this is the kind of change we'd bump semantic versioning for. Does anyone disagree? 

I'm ok to leave "BREAKING" or whatever in the release notes, I'm just loathe to do anything that causes all our downstream crate users to be on an old version.

**Review by michaelkirk — APPROVED** [review]
The code all LGTM, but I want to leave it unmerged until someone else weighs in on not bumping semver for this change.

File: geo/src/algorithm/lines_iter.rs
Comment (callpraths) [reply geo/src/algorithm/lines_iter.rs:9]: I was also conflicted about `Point` and `MultiPoint`. There is the better ergonomics of being able to iterate over the lines of a `Geometry`, but clients lose some help from compiler in catching cases where they probably don't want to iterate over lines in a `Point`.

I'd vote to not add it right now.

File: geo/src/algorithm/lines_iter.rs
Comment (callpraths) [reply geo/src/algorithm/lines_iter.rs:147]: Added TODO

File: geo/src/algorithm/lines_iter.rs
Comment (callpraths) [reply geo/src/algorithm/lines_iter.rs:136]: Added TODO.

Will keep an eye on #741 and #758 and update this once those are merged.

Comment (callpraths) [reply geo/src/algorithm/lines_iter.rs:136]: To double check, we'll still need #741 to be merged, right? Because #758 doesn't actually bump MSRV, so our clients could still be on pre-2021 rustc.

Comment (michaelkirk) [reply geo/src/algorithm/lines_iter.rs:136]: FYI #741 is building now. I expect it to be merged imminently. 

Sorry for the rigamarole as all the trains come into the station at once...

File: geo/src/algorithm/lines_iter.rs
Comment (frewsxcv) [reply geo/src/algorithm/lines_iter.rs:9]: Sounds great, thanks for the feedback!

File: geo/src/algorithm/lines_iter.rs
Comment (callpraths) [reply geo/src/algorithm/lines_iter.rs:136]: No worries. I'm happy to have converted a TODO to a to-done ;)

File: geo-types/CHANGES.md
Comment (callpraths) [reply geo-types/CHANGES.md:5]: I don't have any context on downstream consumers yet, so I don't have a clear opinion here.

I think this falls in the "Possibly-breaking change" per [cargo book](https://doc.rust-lang.org/cargo/reference/semver.html) and in unstable packages, it's OK to bump just the patch version for these changes. The order of returned lines isn't part of an explicit contract, so someone depending on that order to *not* conform to the winding order is unlikely.

Let me know if I can help make this decision (perhaps start a thread on the channel?)

Comment (michaelkirk) [reply geo-types/CHANGES.md:5]: In the interest of a certain amount of efficiency, I don't expect to get an explicit "OK" on all things. Typically when I'm about to do something potentially controversial, my approach is to make a proposal in the code review and solicit input. If no one complains (we're all async, so I wait a couple days) I go forward. 

If it were something that seemed potentially *very* controversial, I might wait for longer or for a more explicit OK from a larger group of people.

Since no one has yet replied to the contrary, I estimate that it's reasonable to move forward without bumping major version at this point.

**Comment by callpraths** [comment]
Thank you for a great first PR experience!

**Pull request merged**
